### PR TITLE
JS: adding type tracking in for portal exit nodes

### PR DESCRIPTION
**Pull request by emarteca**
This adds type tracking to the portal exit nodes. The initial motivation was to be able to find exit nodes where the portal was passed as an argument to a function (which matched to the NpmPackagePortal), but now this type tracking has been added to all the types of portals.

Note: no type tracking has been added to portal entry nodes, since they are not necessarily SourceNodes.

### CODE REVIEW

**Comment by emarteca** [comment]
@xiemaisi

File: javascript/ql/test/library-tests/Portals/src/typeTracking/prototypeFlow.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// example adapted from test cases found in github projects
```
Comment (xiemaisi) [inline javascript/ql/test/library-tests/Portals/src/typeTracking/prototypeFlow.js]: For examples based on third-party code, we generally provide a reference for where we found it and what the license of the code is. That seems like overkill in this case. Could you perhaps replace this with a synthetic test that tests the same functionality, but is _not_ derived from other code?

Comment (emarteca) [reply javascript/ql/test/library-tests/Portals/src/typeTracking/prototypeFlow.js]: Good point, thanks! New commit makes the examples smaller and no longer taken from other github projects.

File: javascript/ql/test/library-tests/Portals/src/typeTracking/simpleField.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// new portal exit node (i.e. caught with typetracking) annotated below
+
+var i = require('./index').NP;
+var np = new i();
+
+np.argFunc( input => {
+	
+	input.on('arg', x => {
+	input.emit( { x: "hello", y: input.field}); // this is a portal exit node, since input is an
```
Comment (xiemaisi) [inline javascript/ql/test/library-tests/Portals/src/typeTracking/simpleField.js]: Which expression does "this" refer to? And can you briefly explain why the portal exit node can only be identified with type tracking?

Comment (emarteca) [reply javascript/ql/test/library-tests/Portals/src/typeTracking/simpleField.js]: Thanks, good catch! I meant the "input.field" expression, but it was actually caught with the original portal exit node tracking. My bad.

New commit removes this test.

**Comment by xiemaisi** [comment]
Tests are failing.

**Comment by xiemaisi** [comment]
On second thought, maybe we should do things a little differently. I think it would be better to keep the exit nodes as they are (and not type-track them), but instead to use type-tracking when we compute one exit node from another. For instance, at the moment `MemberPortal::reads` uses purely local data flow to find property reads on exit nodes of the base portal. Instead, we should use type tracking to identify those property reads.

Implementing this shouldn't be very difficult: we just need one predicate that type-tracks portal exit nodes  (it can be an internal helper predicate in `Portals.qll` for now), and then use that wherever we directly use exit nodes at the moment. We probably want something similar for type-backtracking portal entry nodes, again not changing the definition of `getAnEntryNode` but rather its uses.

Does that make sense?

**Comment by emarteca** [comment]
> On second thought, maybe we should do things a little differently. I think it would be better to keep the exit nodes as they are (and not type-track them), but instead to use type-tracking when we compute one exit node from another. For instance, at the moment `MemberPortal::reads` uses purely local data flow to find property reads on exit nodes of the base portal. Instead, we should use type tracking to identify those property reads.
> 
> Implementing this shouldn't be very difficult: we just need one predicate that type-tracks portal exit nodes (it can be an internal helper predicate in `Portals.qll` for now), and then use that wherever we directly use exit nodes at the moment. We probably want something similar for type-backtracking portal entry nodes, again not changing the definition of `getAnEntryNode` but rather its uses.
> 
> Does that make sense?

Good idea, thanks! New commit implements it this way, and updates the new test to be relevant (since the new exit node considered with the original approach is no longer caught here now that we have removed typetracking from the NpmPackagePortal getAnExitNode; the new test has nonlocal dataflow on a property read).

File: javascript/ql/test/library-tests/Portals/src/typeTracking/nonlocalDataflowRead.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+var n = require('./index').NP;
+var np = new n();
+
+function getPortalReturn() {
+	return np.createNP();
+}
+
+function main() {
+  getPortalReturn().argFunc(); // here we have a nonlocal property access
+}							   // now, with typetracking, this is considered a portal exit node
```
Comment (xiemaisi) [inline javascript/ql/test/library-tests/Portals/src/typeTracking/nonlocalDataflowRead.js]: This comment makes sense as part of the PR, but not as part of checked-in code. I'd suggest omitting it (but keeping the one on the line above).

File: javascript/ql/test/library-tests/Portals/src/typeTracking/nonlocalDataflowRead.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// new portal exit node (i.e. caught with typetracking) annotated below
```
Comment (xiemaisi) [inline javascript/ql/test/library-tests/Portals/src/typeTracking/nonlocalDataflowRead.js]: This won't be "new" any more soon. I don't think this comment is needed.

**Pull request merged**